### DISCUSSION TITLE
How to impose weak spring boundary condition in MOOSE

### DISCUSSION THREAD

**Discussion by BoZeng1997** [idaholab/moose#22022] [Q&A Modules: Solid mechanics] [answered]
Hi,

I want to know how to impose weak spring boundary condition like in ANSYS when using MOOSE. My simulated object swells and I want to have the thermal expansion be almost free on most surfaces. I am trying not to assume how it should swell by fixing the rigid body motion by `DirichletBC`. I am currently considering Robin BC so that traction will be proportional to the displacement. $\frac{\partial u}{\partial n}=ku$. `k` will be a small negative number. But simulation did not converged with `CoupledVarNeumannBC`. 

1. Is Robin BC the right BC to impose weak spring BC in MOOSE? If not, does MOOSE have one?
2. Is `CoupledVarNeumannBC` the right MOOSE object to use for Robin BC for displacement field? If not, does MOOSE have one?
3. Does MOOSE have a warning option for rigid body motion? Sometimes for simple problems, the simulation converged to a reasonable result even the displacement constraints did not forbid rigid body motion.

**Comment by hugary1995** [comment, marked as answer]
1 & 2, a spring BC looks like `stress . normal = k u`. So yes, the `CoupledVarNeumannBC` should work, you may need to repeat it three times for the three components though (twice in 2D, obviously).
3 is a tricky one. You can test for null space by looking at the singular values of your matrix. To do that, first reduce the size of your problem, then run it with `-pc_type svd`. If the minimum singular value is near zero then you definitely have a null space. Whether that null space corresponds to a rigid body mode is at your discretion.
